### PR TITLE
DAOS-16556 client: call fstat() before mmap() to update file status in kernel

### PR DESCRIPTION
**Pull request by wiliamhuang**
In the compatible mode of libpil4dfs, write() is done in user mode. We need to let kernel to update file status before calling mmap().

Features: pil4dfs
Skip-func-test-leap15: false
Skip-func-vm-test-leap15: false

Required-githooks: true
Skipped-githooks: codespell

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by wiliamhuang** [comment]
All tests in CI (including Functional on Leap 15.5) passed.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 1]
```diff
@@ -6302,6 +6302,14 @@ new_mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset)
 
 	atomic_fetch_add_relaxed(&num_mmap, 1);
 
+	if ((fd < FD_FILE_BASE) && (fd_directed >= FD_FILE_BASE) && d_compatible_mode) {
+		/* DAOS-14494: Force the kernel to update the size before mapping. */
```
Comment (ashleypittman) [inline src/client/dfuse/pil4dfs/int_dfs.c:6306]: I've linked DAOS-14494 and DAOS-16556 in Jira.

**Review by ashleypittman — APPROVED** [review]

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:6306]: Sounds good. Thank you very much!

**Comment by wiliamhuang** [comment]
@mchaarawi Could you please review this PR when you have time? It is a small one. Thank you very much!

**Comment by wiliamhuang** [comment]
@mchaarawi Shall we back port this PR to 2.6? Thank you!

**Comment by mchaarawi** [comment]
> @mchaarawi Shall we back port this PR to 2.6? Thank you!

yes makes sense to do that

**Pull request merged**